src/receive_buffer.cpp

```diff
@@ -41,44 +41,66 @@ namespace libtorrent {
 		}
 	}
 
-int receive_buffer::max_receive()
+int receive_buffer::max_receive() const
 {
-	int max = packet_bytes_remaining();
-	if (m_recv_pos >= m_soft_packet_size) m_soft_packet_size = 0;
-	if (m_soft_packet_size && max > m_soft_packet_size - m_recv_pos)
-		max = m_soft_packet_size - m_recv_pos;
-	return max;
+	return int(m_recv_buffer.size() - m_recv_end);
 }
 
 boost::asio::mutable_buffer receive_buffer::reserve(int size)
 {
 	TORRENT_ASSERT(size > 0);
 	TORRENT_ASSERT(m_recv_pos >= 0);
-	// this is unintuitive, but we used to use m_recv_pos in this function when
-	// we should have used m_recv_end. perhaps they always happen to be equal
-	TORRENT_ASSERT(m_recv_pos == m_recv_end);
 
 	// normalize() must be called before receiving more data
 	TORRENT_ASSERT(m_recv_start == 0);
 
-	m_recv_buffer.resize(m_recv_end + size);
+	if (m_recv_buffer.size() < m_recv_end + size)
+	{
+		// TODO: 4 if the message size is larger still, allocate enough space to
+		// fit the whole message
+		buffer new_buffer(m_recv_end + size
+			, aux::array_view<char const>(m_recv_buffer.ptr(), m_recv_end));
+		m_recv_buffer = std::move(new_buffer);
+
+		// since we just increased the size of the buffer, reset the watermark to
+		// start at our new size (avoid flapping the buffer size)
+		m_watermark = sliding_average<20>();
+	}
+
 	return boost::asio::buffer(&m_recv_buffer[0] + m_recv_end, size);
 }
 
+void receive_buffer::grow(int const limit)
+{
+	int const current_size = int(m_recv_buffer.size());
+	TORRENT_ASSERT(current_size < std::numeric_limits<int>::max() / 3);
+
+	// first grow to one piece message, then grow by 50% each time
+	int const new_size = std::min((current_size < m_packet_size)
+		? m_packet_size : current_size * 3 / 2, limit);
```

Comment:
I don't think this is the place the enforce that limit though. I suspect that if you do get into a situation where `m_packet_size > limit` this code will cause the connection to stall. The limit on m_packet_size needs to be enforced further up the call stack where we can handle it gracefully.

The bt packet size is already capped to 1MB by `bt_peer_connection`. The one hole I'm aware of is when authenticated crypto is in use we don't cap the crypto packet size. Running over loopback I've seen crypto packets well over 1MB so the limit will need to be enforced on both the send and receive side.